hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAccountWipeHandler.java

```diff
@@ -54,8 +84,123 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
+        // Wrapping the serial nums this way de-duplicates the serial nums:
+        final var nftSerialNums = new ArrayList<>(new LinkedHashSet<>(op.serialNumbers()));
+
+        // Validate the semantics of the transaction
+        final var validated = validateSemantics(
+                accountId,
+                tokenId,
+                fungibleBurnCount,
+                nftSerialNums,
+                accountStore,
+                tokenStore,
+                tokenRelStore,
+                expiryValidator);
+        final var acct = validated.account();
+        final var token = validated.token();
+
+        final long newTotalSupply;
+        final long newAccountBalance;
+        final Account.Builder updatedAcctBuilder = validated.account().copyBuilder();
+        if (token.tokenType() == TokenType.FUNGIBLE_COMMON) {
+            // Check that the new total supply will not be negative
+            newTotalSupply = token.totalSupply() - fungibleBurnCount;
+            validateTrue(newTotalSupply >= 0, INVALID_WIPING_AMOUNT);
+
+            // Check that the new token balance will not be negative
+            newAccountBalance = validated.accountTokenRel().balance() - fungibleBurnCount;
+            validateTrue(newAccountBalance >= 0, INVALID_WIPING_AMOUNT);
+        } else {
+            // Check that the new total supply will not be negative
+            newTotalSupply = token.totalSupply() - nftSerialNums.size();
+            validateTrue(newTotalSupply >= 0, INVALID_WIPING_AMOUNT);
+
+            // Load and validate the nfts
+            for (final Long nftSerial : nftSerialNums) {
+                final var nft = nftStore.get(tokenId, nftSerial);
+                validateTrue(nft != null, INVALID_NFT_ID);
+
+                final var nftOwner = nft.ownerNumber();
+                validateTrue(nftOwner == accountId.accountNum(), ACCOUNT_DOES_NOT_OWN_WIPED_NFT);
+            }
+
+            // Check that the new token balance will not be negative
+            newAccountBalance = validated.accountTokenRel().balance() - nftSerialNums.size();
+            validateTrue(newAccountBalance >= 0, INVALID_WIPING_AMOUNT);
+
+            // Update the NFT count for the account
+            updatedAcctBuilder.numberOwnedNfts(acct.numberOwnedNfts() - nftSerialNums.size());
+
+            // Remove the NFTs
+            nftSerialNums.forEach(serialNum -> nftStore.remove(tokenId, serialNum));
+        }
+
+        // Finally, record all the changes
+        if (newAccountBalance == 0) {
+            updatedAcctBuilder.numberPositiveBalances(acct.numberPositiveBalances() - 1);
+        }
+        accountStore.put(updatedAcctBuilder.build());
+        tokenStore.put(token.copyBuilder().totalSupply(newTotalSupply).build());
+        tokenRelStore.put(validated
+                .accountTokenRel()
+                .copyBuilder()
+                .balance(newAccountBalance)
+                .build());
```

Comment:
We need to add a comment that the records will be built in `finalize` method, so that we keep track of all changes for records

Reply:
Done